From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shelley Vohr <shelley.vohr@gmail.com>
Date: Mon, 5 Aug 2024 13:26:02 +0000
Subject: Fix potential draggable region crash when no MainFrameImpl

This CL fixes a crash that can occur when SetSupportsDraggableRegions
is called with `true` and there is no MainFrameImpl. When MainFrameImpl
is nullptr, logic currently correctly returns early, but supports_draggable_regions_ is set before that happens. As a result, when SupportsDraggableRegions() is called, it will return true, and thus LocalFrameView::UpdateDocumentDraggableRegions() will call DraggableRegionsChanged(). This will trigger a crash in WebViewImpl::DraggableRegionsChanged(), as it assumes that MainFrameImpl is not null.

Chrome likely does not hit this crash because it only collects app regions for Window Controls Overlay & Borderless apps, whereas Electron enables draggable regions for more frames.

Change-Id: I80988734b38f143851043b36914fb5eff6135fd7
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5756619
Reviewed-by: Amanda Baker <ambake@microsoft.com>
Reviewed-by: Dmitry Gozman <dgozman@chromium.org>
Commit-Queue: Nico Weber <thakis@chromium.org>
Reviewed-by: Nico Weber <thakis@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1337265}

diff --git a/third_party/blink/renderer/core/exported/web_view_impl.cc b/third_party/blink/renderer/core/exported/web_view_impl.cc
index 948cf94e2e4af0bbbf1f9c2322d00075bdaca0b2..dff4f7ee7163648eae4513e39cf52b9e7e7a0035 100644
--- a/third_party/blink/renderer/core/exported/web_view_impl.cc
+++ b/third_party/blink/renderer/core/exported/web_view_impl.cc
@@ -4095,6 +4095,10 @@ bool WebViewImpl::SupportsDraggableRegions() {
 }
 
 void WebViewImpl::DraggableRegionsChanged() {
+  if (!MainFrameImpl()) {
+    return;
+  }
+
   WebVector<WebDraggableRegion> web_regions =
       MainFrameImpl()->GetDocument().DraggableRegions();
 
